File: src/y2025/day04.rs
//! <https://adventofcode.com/2025/day/4>
use {
    crate::{
        framework::{aoc, AdventOfCode, ParseError},
        geometric::{neighbors8, Dim2},
    },
    rustc_data_structures::fx::{FxHashMap, FxHashSet, FxHasher},
    std::{
        collections::{HashMap, HashSet},
        hash::BuildHasherDefault,
        mem::swap,
    },
};

#[derive(Clone, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub struct Puzzle {
    line: Vec<Vec<bool>>,
}

mod parser {
    use winnow::{
        ascii::newline,
        combinator::{repeat, separated},
        token::one_of,
        ModalResult, Parser,
    };

    fn parse_line(s: &mut &str) -> ModalResult<Vec<bool>> {
        repeat(1.., one_of(['.', '@']).map(|c: char| c == '@')).parse_next(s)
    }
    pub fn parse(s: &mut &str) -> ModalResult<Vec<Vec<bool>>> {
        separated(1.., parse_line, newline).parse_next(s)
    }
}

#[aoc(2025, 4)]
impl AdventOfCode for Puzzle {
    fn prepare(&mut self, mut input: &str) -> Result<(), ParseError> {
        self.line = parser::parse(&mut input)?;
        Ok(())
    }
    fn part1(&mut self) -> Self::Output1 {
        let mut found: usize = 0;
        let height = self.line.len();
        let width = self.line[0].len();
        for (i, l) in self.line.iter().enumerate() {
            for (j, b) in l.iter().enumerate() {
                if *b
                    && neighbors8(i, j, height, width)
                        .iter()
                        .filter(|(i, j)| self.line[*i][*j])
                        .count()
                        < 4
                {
                    found += 1;
                }
            }
        }
        found
    }
    fn part2(&mut self) -> Self::Output2 {
        let height = self.line.len();
        let width = self.line[0].len();
        let num_rolls = self
            .line
            .iter()
            .map(|l| l.iter().filter(|b| **b).count())
            .sum();
        let mut roll_id: FxHashMap<Dim2<usize>, usize> =
            HashMap::<_, _, BuildHasherDefault<FxHasher>>::default();
        let mut propagate = vec![Vec::<usize>::new(); num_rolls];
        let mut count = vec![1_u8; num_rolls]; // 0 for dead
        for (i, l) in self.line.iter().enumerate() {
            for (j, b) in l.iter().enumerate() {
                if *b {
                    let n = roll_id.len();
                    let p_id = *roll_id.entry((i, j)).or_insert(n);
                    neighbors8(i, j, height, width)
                        .iter()
                        .filter(|q| self.line[q.0][q.1])
                        .for_each(|q| {
                            let n = roll_id.len();
                            let q_id = *roll_id.entry(*q).or_insert(n);
                            propagate[p_id].push(q_id);
                            count[p_id] += 1;
                        });
                }
            }
        }
        let mut removables: FxHashSet<usize> =
            HashSet::<_, BuildHasherDefault<FxHasher>>::default();
        for (id, c) in count.iter().enumerate() {
            if *c < 4 + 1 {
                removables.insert(id);
            }
        }
        let mut next: FxHashSet<usize> = HashSet::<_, BuildHasherDefault<FxHasher>>::default();
        let mut num_deads = 0;
        while !removables.is_empty() {
            next.clear();
            for p in removables.iter() {
                if count[*p] > 0 {
                    count[*p] = 0;
                    num_deads += 1;
                    for q in propagate[*p].iter() {
                        if count[*q] > 0 {
                            count[*q] -= 1;
                            if count[*q] < 4 + 1 {
                                next.insert(*q);
                            }
                        }
                    }
                }
            }
            swap(&mut next, &mut removables);
        }
        num_deads
    }
}

File: Swift/Sources/Y2025/day04.swift
import Parsing
import Utils

private func part1(_ grid: [[Bool]]) -> Int {
    var found = 0
    let height = grid.count
    let width = grid.first?.count ?? 0
    let boundary = Pos(y: height - 1, x: width - 1)
    for i in grid.indices {
        for j in grid[i].indices {
            if grid[i][j] {
                var count = 0
                for p in Pos(y: i, x: j).neighbors8(bound: boundary) {
                    if grid[p.y][p.x] {
                        count += 1
                    }
                }
                found += count < 4 ? 1 : 0

            }
        }
    }
    return found
}

private func part2(_ grid: [[Bool]]) -> Int {
    let height = grid.count
    let width = grid.first?.count ?? 0
    let boundary = Pos(y: height - 1, x: width - 1)
    let num_rolls =
        grid
        .map { $0.filter { $0 }.count }
        .reduce(0, +)
    var roll_id: [Pos: Int] = [:]
    var propagate: [[Int]] = Array(repeating: [Int](), count: num_rolls)
    var count: [Int] = Array(repeating: 1, count: num_rolls)
    for i in grid.indices {
        for j in grid[i].indices {
            let p = Pos(y: i, x: j)
            if grid[i][j] {
                roll_id[p] = roll_id[p] ?? roll_id.count
                for q in p.neighbors8(bound: boundary) {
                    if grid[q.y][q.x] {
                        roll_id[q] = roll_id[q] ?? roll_id.count
                        propagate[roll_id[p]!].append(roll_id[q]!)
                        count[roll_id[p]!] += 1
                    }
                }
            }
        }
    }
    var removables: Set<Int> = Set()
    for i in count.indices {
        let c = count[i]
        if c < 5 {
            removables.insert(i)
        }
    }
    var next: Set<Int> = Set()
    var numDeads = 0
    while !removables.isEmpty {
        next.removeAll()
        for p in removables {
            if count[p] > 0 {
                count[p] = 0
                numDeads += 1
                for q in propagate[p] {
                    if count[q] > 0 {
                        count[q] -= 1
                        if count[q] < 5 {
                            next.insert(q)
                        }
                    }
                }
            }
        }
        (next, removables) = (removables, next)
    }
    return numDeads
}

public func day04(_ data: String) {
    let line_parser: some Parser<Substring, [Bool]> = Many {
        First<Substring>()
            .filter { $0 == "@" || $0 == "." }
            .map { $0 == "@" }
    } terminator: {
        "\n"
    }
    let parser: some Parser<Substring, [[Bool]]> = Many {
        line_parser
    }
    do {
        let grid = try parser.parse(data)
        let sum1 = part1(grid)
        let sum2 = part2(grid)
        print("Part 1: \(sum1)")
        print("Part 2: \(sum2)")
    } catch {
        print(error)
        fatalError()
    }
}
